### ISSUE TITLE
Cannot define different excluded styles for different color rules

### ISSUE THREAD

**Issue by Tankster** [Ekopalypse/EnhanceAnyLexer#6]
Hi

So I currently only have two rules in place on top of my UDL
1. I want all path/filenames to be a certain color (gray)
2. I want any text that starts with a $ to be a certain color. (yellow)

Also, for context, in my UDL, all text within quotations should be green

When it comes to excluding styles, I want 1 to exclude all styles, but I want 2 to exclude all except for within quotations (This is style 16 for me, as it is the only Delimiter I have defined in the UDL).
i.e. I want any usage of $keyword to show up yellow, despite being within quotations, but I don't want folder paths/filenames to show up gray within quotations.

The problem is that I cannot seem to apply this configuration. It either ends up with both rule 1 and 2 being colored within quotations, or I end up turning off rule 1 completely. 
Is there a way to make different sets of excludes for each set of rules?

**Comment by Tankster** [comment]
Thanks for the suggestion. I might move the delimiter to the rules for now.

One issue with it is that I wanted this UDL to be contained in one .xml file as far as possible so I can share it with colleagues without telling them to also install an extra plugin for the full intended functionality.
I feel that the delimiter rule is much more important than the two rules I've put in the plugins (they are more luxury really), but for my own use - since I already have the plugin - I might try that.

My initial hope was that you would be able to apply the _excluded_styles_ styles multiple times, so that an _excluded_styles_ line would apply to each regex above it, up until either the [udl] tag or until the previous _excluded_styles_ line. That felt intuitive to me, but I don't know if that covers all potential use cases

**Comment by Ekopalypse** [comment]
If `$` is unique you might be able to make it work with UDL only by defining a keyword group

![image](https://github.com/Ekopalypse/EnhanceAnyLexer/assets/47723516/1a4b4624-cde4-4c1d-beb5-34e37e1a62e3)

and allow nesting from delimiter

![image](https://github.com/Ekopalypse/EnhanceAnyLexer/assets/47723516/e9b3b315-b2b7-4e64-b4eb-ba9149025082)

![image](https://github.com/Ekopalypse/EnhanceAnyLexer/assets/47723516/295ddb4f-73b5-4fbf-bb4c-c568a26acbf6)

**Comment by Tankster** [comment]
The problem is that the usage of $ is often in combination with paths. for example:
![image](https://github.com/Ekopalypse/EnhanceAnyLexer/assets/6097710/9d23c2a3-4829-4d5d-9ad2-39b432b1a1f2)

and since the plugin colors everything **after** the udl is done (afaik), it will paint over all paths with the gray color, leading to this instead:
![image](https://github.com/Ekopalypse/EnhanceAnyLexer/assets/6097710/65a4bb24-f954-496f-921e-a37e50904446)

by the way, these are the regex rules I have for $ and path. 

```
0xdaaaaa=(\w{1}:\/)|([.]{1,2}\/)+|((\w+\/)+(\w*))+|(\w*[.]\w+)?
0x00ffff=[$]\w+
```

It's not a very good regex because https://regexr.com/ even gives me an infinite match warning, and it technically doesn't match the whole string at once, but rather in parts, but it works sooo...  
the path is made to catch things like ./ and ../ too

**Comment by Ekopalypse** [comment]
Sorry for the confusion, my reply was in reference to your note about making this available to your colleagues who don't necessarily have additional plugins installed. But the keyword prefix trick would of course mean that everything after the dollar sign would be colored yellow.

![image](https://github.com/Ekopalypse/EnhanceAnyLexer/assets/47723516/07885aba-2022-4bc5-b0a7-ff4f76b3aa18)

I thought about adding a "whitelist" option per regex, using a list of style ids enclosed in square brackets, e.g. `0x0ffff[3,4,5,6]=\$\w+`
which would mean that this regex is applied to areas already styled by these IDs. This does not mean that all other ids are automatically excluded, they still need to be expressed by the excluded_styles keyword.
Which means that this whitelist only makes sense if there is an excluded_styles.

As for your regex, what matters is that it works, right. As far as I can see, you are only processing Unix-style paths, i.e. with slashes as path separators. If that's the case, you could simplify it by using a regex that expresses what is not allowed, like `(?:[a-zA-Z]:)*(?:[\/\\][^\*\"\<\>\:\|\?\s]+)`.

**Comment by Tankster** [comment]
> Sorry for the confusion, my reply was in reference to your note about making this available to your colleagues who don't necessarily have additional plugins installed.

Aha I see, yes that is tempting.

>  But the keyword prefix trick would of course mean that everything after the dollar sign would be colored yellow.

Yes that is indeed another issue that shows up with that solution. For this reason I haven't added any support for $keyword or paths in the .xml file I shared with colleagues.

> I thought about adding a "whitelist" option per regex, using a list of style ids enclosed in square brackets, e.g. `0x0ffff[3,4,5,6]=\$\w+` which would mean that this regex is additionally applied to areas already styled by these IDs.

That's probably a nice solution, and it should be backwards compatible too. 

>  Which means that this whitelist only makes sense if there is an excluded_styles.

The way I saw it when installing the plugin - and also considering the plugins name - it felt reasonable to assume that the default behavior of the plugin wasn't to completely override all the UDL rules, but rather 'enhance' them by being able to use regex for formatting. 
What I mean by this is that starting the config with a full list of all IDs in the excluded_styles line seems like a reasonable default starting point for just enhancing an UDL. If you then manually list IDs in a whitelist config, you're basically slowly overriding your UDL more and more instead of just enhancing it, which is reasonable when using advanced configs imo.

> As for your regex, what matters is that it works, right. As far as I can see, you are only processing Unix-style paths, i.e. with slashes as path separators. If that's the case, you could simplify it by using a regex that expresses what is not allowed, like `(?:[a-zA-Z]:)*(?:[\/\\][^\*\"\<\>\:\|\?\s]+)`.

Thanks. I haven't used non-capturing groups before so not sure how to utilize it. Maybe I could tweak that one to include paths that start with ../ or ./ too. 
Anyway, I suppose that doesn't relate to this issue really hehe.

**Comment by Ekopalypse** [comment]
Will be implemented with version 1.2.0, which is currently marked as draft for a test period of one week.

**Comment by Ekopalypse** [comment]
implemented and released
